platform/windows/adapter/mellanox/mellanox.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2017 Microsoft. All rights reserved.
+// MIT License
+
+package mellanox
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-container-networking/log"
+)
+
+const (
+	// Search string to find adapter having Mellanox in description
+	mellanoxSearchString = "*Mellanox*"
+
+	// PriorityVlanTag reg key for adapter
+	priorityVLANTagIdentifier = "*PriorityVLANTag"
+
+	// Registry key Path Prefix
+	registryKeyPrefix = "HKLM:\\System\\CurrentControlSet\\Control\\Class\\"
+)
+
+var (
+	errorMellanoxAdapterNotFound = fmt.Errorf("no network adapter found with %s in description", mellanoxSearchString)
+	errorMellanoxDeviceNotFound  = fmt.Errorf("no network device found with %s in description", mellanoxSearchString)
+	errorPowershellNotFound      = fmt.Errorf("failed to find powershell executable")
+)
+
+type Mellanox interface {
+	// GetAdapter returns name of Mellanox adapter if found
+	// Must return errorMellanoxAdapterNotFound if adapter is not found or adapter name empty
+	GetAdapaterName() (string, error)
+
+	// Get PriorityVLANTag returns PriorityVLANTag value for Mellanox Adapter (both version 3 and version 4)
+	GetPriorityVLANTag() (int, error)
+
+	// Set Mellanox adapter's PriorityVLANTag value to desired value if adapter exists
+	SetPriorityVLANTag(int) error
+}
+
+type MellanoxImpl struct{}
```

Comment:
`Impl` is a bit of a naming anti-pattern in Go, because it implies things are created for the purpose of implementing an interface, therefore coupling them. Given the package, I think naming this type `Controller` seems appropriate. A few more things:
- The interface should be defined where it's used, not where it's implemented.
- Curious about the file structure. Can this package just be under `mellanox` instead of `platform/windows/adapter/mellanox`?

Reply:
Definitely agree with `Impl`, but I actually think the interface name is the problem here. What does it mean for a set of behavior (i.e. an interface) to be a "Mellanox"? I think instead, something like "NetworkAdapter" is better... things that claim to be an network adapter should be able to tell you their name and get/set priority VLAN tags. This struct should be "Mellanox" or "MellanoxAdapter".

Reply:
thanks @ramiro-gamarra and @timraymond for this feedback... makes sense.. so I will name Interface as NetworkAdapter and will move it under adapter/networkadapter.go and will rename implementation as Mellanox instead of MellanoxImpl which will make more sense.... 
and the reason I have file structure as windows/adapter... since I am specifically talking about windows here.. not sure if you have anyother suggestions for it....